cloud-engineering-with-gcp/python/content/lab-02/README.md

```diff
@@ -0,0 +1,199 @@
+    entity="allUsers"
+)
+```
+
+## Step 3 &mdash; Upload your Objects
+
+Now we need to upload the files that comprise our website so we can view them. Because Pulumi uses real programming languages, we can use constructs like `for` loops. Let's use a `for` loop to iterate over the files in the `wwwroot` directory in this repo and upload them using the `BucketObject` resource.
+
+First, we need to import the `os` Python library. Note that a major advantage of Pulumi's design of using real programming languages is that we can make use of both standard libraries and external packages when defining our infrastructure.
+
+Add the following statement near the top of your `__main__.py` near your other imports:
+
+```python
+import os
+```
+
+```python
+for file in ["404.html", "index.html"]:
+    filepath = os.path.join("../wwwroot", file)
+    gcp.storage.BucketObject(
+        file,
+        bucket=bucket.name,
+        name=file,
+        source=pulumi.FileAsset(filepath),
+        opts=pulumi.ResourceOptions(depends_on=[acl])
+    )
+```
+
+Notice the use of `depends_on`. This tells Pulumi that our `BucketObjects` should not be created until our `DefaultObjectAccessControl` has been fully provisioned. The `depends_on` is necessary because there's no _explicit_ dependency between the files and the ACL (i.e., there's no output from the `DefaultObjectAccessControl` resource passed to the `BucketObject` inputs). If we didn't explicitly specify `depends_on`, our files may get uploaded before the default ACL is applied, and our files would not get created with the right default permissions.
+
+## Step 4 &mdash; Run `pulumi up`
+
+Now that we've defined our infrastructure, we use the Pulumi CLI to create the resources we've defined.
+
+Run `pulumi up` in your project directory. You should see something like this:
+
+```bash
+$ pulumi up
+Previewing update (dev)
+
+View Live: https://app.pulumi.com/jaxxstorm/static_site/dev/previews/b5ce861f-16ac-4930-ada4-1e9547e02eb1
+
+     Type                                       Name             Plan
+ +   pulumi:pulumi:Stack                        static_site-dev  create
+ +   ├─ gcp:storage:Bucket                      website          create
+ +   ├─ gcp:storage:DefaultObjectAccessControl  website          create
+ +   ├─ gcp:storage:BucketObject                404.html         create
+ +   └─ gcp:storage:BucketObject                index.html       create
+
+Resources:
+    + 5 to create
+
+Do you want to perform this update?  [Use arrows to move, enter to select, type to filter]
+  yes
+> no
+```
+
+You can examine the details of the resources that will be created. When you're happy, move the arrow to `yes` and watch as Pulumi creates your resources!
+
+## Step 5 &mdash; Export the Bucket URL
+
+Our final step is to build our static site URL and add it as a [stack output](https://www.pulumi.com/learn/building-with-pulumi/stack-outputs/).
+
+First, we assemble the output's value using `pulumi.Output.concat`:
```

Comment:
People are going to get very lost with the introduction of Outputs here (Python has a lot of issues with them in general), which is why we don't bring in this kind of concat situation in the general beginner workshops. I'd recommend spending time here explaining it.

Note that we do not cover inputs and outputs in the fundamentals pathways for this reason and based off of a lot of feedback.

Reply:
This is explained just below.